gax-java/gax/src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java

```diff
@@ -232,6 +225,28 @@ public boolean shouldRetry(TimedAttemptSettings nextAttemptSettings) {
     return true;
   }
 
+  // For non-LRO RPCs, do not attempt to retry if the totalTime spend is over
+  // the totalTimeout (timeout is in the past) or at the totalTimeout (timeout
+  // will occur immediately). For any other value, the deadlineScheduler will
+  // terminate in the future (even if the timeout is small).
+  @InternalApi
+  protected boolean shouldRPCTerminate(
+      long totalTimeSpentNanos, long totalTimeoutNanos, Duration rpcTimeout) {
+    // For RPC Retry Requests, the rpcTimeout in `createNextAttempt()` is calculated with
+    // millisecond precision. It is possible that there are a few nanoseconds before the
+    // totalTimeout (which would allow for a retry), but the RPC Timeout would be 0
+    // (as 999999ns or less is calculated as 0ms). For an RPC Timeout value of 0,
+    // it would mean that the Callable does not set an explicit timeout and that the request
+    // would continue until it either receives a response or an exception from the socket.
+    //
+    // This is *hopefully* a rare case that may only occur when there it is close to the
+    // totalTimeout deadline, but this should try to account for the differences in units
+    // being used for calculations (nanos vs ms). An RPC Timeout may not need nanosecond
+    // precision as it is possible that it would send the request and then immediately times
+    // out.
+    return totalTimeSpentNanos >= totalTimeoutNanos || rpcTimeout.isZero();
+  }
```

Comment:
@blakeli0 This was the section in the retryAlgorithm that I modified. Adding in `rpcTimeout.isZero()` for non-LRO logic. Wanted to double check that this is fine.

Reply:
I need to understand this scenario better, let's discuss it in Wednesday's meeting.